README.md
# ui-search

Copyright (C) 2017-2018 The Open Library Foundation

This software is distributed under the terms of the Apache License,
Version 2.0. See the file "[LICENSE](LICENSE)" for more information.

## Introduction

The Search UI Module, or `ui-search`, is a Stripes UI module used for searching, sorting, filtering and viewing records from [the FOLIO Codex](https://www.openlibraryenvironment.org/archives/411), an aggregation of bibliographic metadata from multiple sources.

## Additional information

Other [modules](https://dev.folio.org/source-code/#client-side).

See project [UISE](https://issues.folio.org/browse/UISE)
at the [FOLIO issue tracker](https://dev.folio.org/guidelines/issue-tracker).

Other FOLIO Developer documentation is at [dev.folio.org](https://dev.folio.org/)


CHANGELOG.md
# Change history for ui-search

## IN PROGRESS

* Bump CI from Node 12 to 16. UISE-154.

## [5.1.0](https://github.com/folio-org/ui-search/tree/v5.1.0) (2022-03-04)
[Full Changelog](https://github.com/folio-org/ui-search/compare/v5.0.0...v5.1.0)
* Hide or fix the permissions related to Codex. UISE-139
* Remove Bigtest tests. UISE-146

## [5.0.0](https://github.com/folio-org/ui-search/tree/v5.0.0) (2021-10-07)
[Full Changelog](https://github.com/folio-org/ui-search/compare/v4.1.0...v5.0.0)
* Increment stripes to v7. UISE-142

## [4.1.0](https://github.com/folio-org/ui-search/tree/v4.1.0) (2021-06-07)
[Full Changelog](https://github.com/folio-org/ui-search/compare/v4.0.0...v4.1.0)
* Correctly provide `notLoadedMessage` in `Search` component
* Update translations

## [4.0.0](https://github.com/folio-org/ui-search/tree/v4.0.0) (2020-03-04)
[Full Changelog](https://github.com/folio-org/ui-search/compare/v3.1.0...v4.0.0)
* Update `stripes` to `v6.0.0`. UISE-136.
* Update `stripes-cli` to `v2.0.0`. UISE-138.
* Update translations

## [3.1.0](https://github.com/folio-org/ui-search/tree/v3.1.0) (2020-10-13)
[Full Changelog](https://github.com/folio-org/ui-search/compare/v3.0.0...v3.1.0)
* Refactor to `miragejs` from `bigtest/mirage`
* Increment `@folio/stripes` to `v5.0.0`, increment `react-intl` to `v5.7.0` UISE-127.
* Update translations

## [3.0.0](https://github.com/folio-org/ui-search/tree/v3.0.0) (2020-06-11)
[Full Changelog](https://github.com/folio-org/ui-search/compare/v2.0.0...v3.0.0)
* Purge `intlShape` in prep for `react-intl` `v4` migration. Refs STRIPES-672.
* Update `stripes` to `v4.0.0`, `stripes-core` to `5.0.0` and added `react-intl` to dev dependencies. UISE-124.
* Add `ui-search.codexsearch` permission. UISE-119.
* Prefer `stripes.actsAs` to the deprecated `stripes.type` in `package.json`. Refs STCOR-148.
* Update translations

## [2.0.0](https://github.com/folio-org/ui-search/tree/v1.10.0) (2020-03-12)
[Full Changelog](https://github.com/folio-org/ui-search/compare/v1.10.0...v2.0.0)

* Update eslint to v6.2.1. Refs UISE-117.
* Update "stripes" to 'v3.0.0', "stripes-core" to '4.0.0' and "react-intl" to '2.9.0'. Refs UISE-122.
* Update of translations

## [1.10.0](https://github.com/folio-org/ui-search/tree/v1.10.0) (2019-12-05)
[Full Changelog](https://github.com/folio-org/ui-search/compare/v1.9.0...v1.10.0)

* Update of translations

## [1.9.0](https://github.com/folio-org/ui-search/tree/v1.9.0) (2019-09-10)
[Full Changelog](https://github.com/folio-org/ui-search/compare/v1.8.0...v1.9.0)

* Update of translations

## [1.8.0](https://github.com/folio-org/ui-search/tree/v1.8.0) (2019-07-24)
[Full Changelog](https://github.com/folio-org/ui-search/compare/v1.7.0...v1.8.0)

* Support translation of filtering option labels. Refs UISE-107.

## [1.7.0](https://github.com/folio-org/ui-search/tree/v1.7.0) (2019-05-09)
[Full Changelog](https://github.com/folio-org/ui-search/compare/v1.6.0...v1.7.0)

* Translate column headers. Fixes UISE-107.
* Gracefully handle missing filters. Refs UISE-110.
* Correctly link to eholdings. Refs UISE-110.

## [1.6.0](https://github.com/folio-org/ui-search/tree/v1.6.0) (2019-03-17)
[Full Changelog](https://github.com/folio-org/ui-search/compare/v1.5.0...v1.6.0)

* Refactor filters. UISE-97.
* Move AppIcon import to `@folio/stripes/core`. Refs STCOM-411.
* Update integration tests to accommodate MCL aria changes. Fixes UISE-102.
* Extract static strings for translation. Fixes UISE-105.
* Add test coverage. UISE-90.

## [1.5.0](https://github.com/folio-org/ui-search/tree/v1.5.0) (2019-01-25)
[Full Changelog](https://github.com/folio-org/ui-search/compare/v1.4.0...v1.5.0)

* Upgrade to stripes v2.0.0.

## [1.4.0](https://github.com/folio-org/ui-search/tree/v1.4.0) (2018-12-13)
[Full Changelog](https://github.com/folio-org/ui-search/compare/v1.3.0...v1.4.0)

* Enable subject search for non-local too
* Clean-up translations etc

## 1.3.0 (https://github.com/folio-org/ui-search/tree/v1.3.0) (2018-10-5)
[Full Changelog](https://github.com/folio-org/ui-search/compare/v1.2.0...v1.3.0)

* Conform references to stripes-components and stripes-smart-components to use new stripes framework.

## 1.2.0 (https://github.com/folio-org/ui-search/tree/v1.2.0) (2018-09-12)
[Full Changelog](https://github.com/folio-org/ui-search/compare/v1.1.0...v1.2.0)

* Update Okapi dependency, `codex`: 2.0
* Change app name to "Codex Search". Applies in the application list in the Stripes menu-bar, within the application itself, and in the top-level route `codexsearch`. Fixes UISE-23.
* New document, [the Codex contract](doc/codex-contract.md), which admittedly sounds like a Dan Brown book. Fixes UISE-28.
* Dummy up the package file to provide an example of app metadata bundle. Fixes UISE-29, towards STCOR-117.
* Display source-type icons instead of 'local', 'kb'. Fixes UISE-6.
* Fix unsupported query to Codex API. Fixes UISE-30.
* Modify set of available indexes based on selected target types. Fixes UISE-34.
* Provide indexes suitable for local-only search. Fixes UISE-31.
* Provide indexes suitable for KB search. Fixes UISE-32.
* Provide indexes suitable when both targets available. Fixes UISE-33.
* Set value of query-index from the `qindex` URL query parameter. Fixes UISE-38.
* Grey out unavailable indexes instead of omitting them entirely. Fixes UISE-40.
* Implement correct filters for local-only targets. Fixes UISE-3.
* Default query for local-only mode is now Codex profile-compliant. Fixes UISE-43.
* When starting the Codex Search UI, do not search for `cql.allRecords=1`. Fixes UISE-41.
* Limit to a single sort-key. Avoids causing problems for mod-ebsco-ekb, and fixes UISE-47.
* Use more-current stripes-components. Refs STRIPES-495.
* Prevent selection of query-index not supported by source. Fixes UISE-46.
* Link from ui-search's KB records into ui-eholdings. Fixes UISE-51, UISE-53, UISE-54 and UISE-55.
* Link from ui-search's local records into ui-inventory. Fixes UISE-50.

src/Search/Search.js
import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import { useIntl } from 'react-intl';

import { AppIcon } from '@folio/stripes/core';

import {
  makeQueryFunction,
  SearchAndSort
} from '@folio/stripes/smart-components';

import { filterState } from '@folio/stripes/components';

import packageInfo from '../../package';

import ViewRecord from '../ViewRecord';
import {
  Filters,
  filterNames,
} from '../Filters';

import {
  redirectParams,
  parseFiltersString,
  availableIndexes,
  filterConfig,
} from './model';

const Search = (props) => {
  const intl = useIntl();

  const updateQIndex = () => {
    const qindex = _.get(props.resources.query, 'qindex', '');
    let indexObject;

    for (const index of availableIndexes) {
      if (index.value === qindex) {
        indexObject = index;
        break;
      }
    }

    if (!indexObject) return;
    if (!indexObject.localOnly) return;

    // Find first available index that is not localOnly
    for (const index of availableIndexes) {
      if (!index.localOnly) {
        props.mutator.query.update({ qindex: index.value });
        return;
      }
    }
  };

  const onFilterChangeHandler = ({ name, values }) => {
    const {
      resources,
      mutator,
    } = props;

    const filters = parseFiltersString(_.get(resources, 'query.filters', ''));

    const updatedActiveFilters = values.length !== 0
      ? {
        ...filters,
        [name]: values,
      }
      : _.omit(filters, [name]);

    const newFilters = [];

    for (const filterName in updatedActiveFilters) {
      if (Object.prototype.hasOwnProperty.call(updatedActiveFilters, filterName)) {
        const filtersString = updatedActiveFilters[filterName].map(filterValue => `${filterName}.${filterValue}`);
        newFilters.push(filtersString);
      }
    }

    mutator.query.update({ filters: newFilters.join(',') });

    if (_.get(updatedActiveFilters, [filterNames.SOURCE], []).join('') === 'local') {
      updateQIndex();
    }
  };

  const onChangeIndex = (e) => {
    const qindex = e.target.value;
    const logger = props.stripes.logger;
    logger.log('action', `changed query-index to '${qindex}'`);
    props.mutator.query.update({ qindex });
  };

  const onSelectRow = (e, record) => {
    const logger = props.stripes.logger;

    const obj = redirectParams(record);
    logger.log('action', `clicked ${record.id}, jumping to '${record.source}' version with obj`, obj);
    if (obj) {
      props.mutator.query.update(obj);
    } else {
      logger.log('action', `unsupported source '${record.source}': doing nothing`);
    }

    return false;
  };

  const renderFilters = (onChange) => {
    return (
      <Filters
        activeFilters={parseFiltersString(_.get(props, 'resources.query.filters', ''))}
        onChange={onChange}
      />
    );
  };

  const resultsFormatter = {
    source: x => (
      <AppIcon
        app={x.source === 'local' ? 'inventory' : 'eholdings'}
        iconKey={x.source === 'local' ? 'instance' : 'app'}
        size="small"
      >
        {x.source === 'local' ? 'Local' : 'KB'}
      </AppIcon>
    ),
    contributor: x => (x.contributor || []).map(y => `'${y.name}'`).join(', '),
  };

  const searchableIndexes = availableIndexes.map(index => ({ ...index }));
  const filters = _.get(props.resources, ['query', 'filters']);
  const filterKeys = filterState(filters);

  if (!filterKeys['source.Local'] || filterKeys['source.Knowledge Base']) {
    for (const index of searchableIndexes) {
      if (index.localOnly) index.disabled = true;
    }
  }

  const translatedSearchableIndexes = searchableIndexes.map(i => {
    const { value, label, ...rest } = i;
    return {
      label: intl.formatMessage({ id: `ui-search.searchableIndexes.${label}` }),
      value,
      ...rest
    };
  });

  return (
    <div data-test-search>
      <SearchAndSort
        packageInfo={packageInfo}
        objectName="record"
        searchableIndexes={translatedSearchableIndexes}
        selectedIndex={_.get(props.resources.query, 'qindex')}
        searchableIndexesPlaceholder={null}
        onChangeIndex={onChangeIndex}
        onFilterChange={onFilterChangeHandler}
        maxSortKeys={1}
        initialResultCount={30}
        resultCountIncrement={30}
        viewRecordComponent={ViewRecord}
        visibleColumns={['source', 'title', 'contributor']}
        columnMapping={{
          source: intl.formatMessage({ id: 'ui-search.header.source' }),
          title: intl.formatMessage({ id: 'ui-search.header.title' }),
          contributor: intl.formatMessage({ id: 'ui-search.header.contributor' }),
        }}
        columnWidths={{ source: '10%', title: '40%', contributor: '50%' }}
        resultsFormatter={resultsFormatter}
        renderFilters={renderFilters}
        onSelectRow={onSelectRow}
        viewRecordPerms="users.item.get"
        disableRecordCreation
        parentResources={props.resources}
        parentMutator={props.mutator}
        notLoadedMessage={intl.formatMessage({ id: 'ui-search.notLoadedMessage' })}
      />
    </div>
  );
};

Search.manifest = Object.freeze({
  resultCount: { initialValue: 30 },
  query: { initialValue: {} },
  records: {
    type: 'okapi',
    records: 'instances',
    path: 'codex-instances',
    recordsRequired: '%{resultCount}',
    perRequest: 30,
    GET: {
      params: {
        query: makeQueryFunction(
          'cql.allRecords=1',
          'title="%{query.query}*" or identifier="%{query.query}*" or contributor="%{query.query}*" or publisher="%{query.query}*"',
          { Title: 'title', Contributor: 'contributor.name' },
          filterConfig,
          true,
        ),
      },
      staticFallback: { params: {} },
    },
  },
});

Search.propTypes = {
  resources: PropTypes.shape({
    query: PropTypes.shape({
      qindex: PropTypes.string,
      query: PropTypes.string,
    }),
  }),
  mutator: PropTypes.shape({
    query: PropTypes.shape({
      update: PropTypes.func.isRequired,
    }),
  }),
  stripes: PropTypes.shape({
    logger: PropTypes.shape({
      log: PropTypes.func.isRequired,
    }).isRequired,
  }).isRequired,
};

export default Search;
